pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -895,14 +1012,198 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
+			if err != nil {
+				klog.ErrorS(err, "GetPodCgroupCpuConfig failed", "pod", pod.Name)
+				return
+			}
+			if errUpdate := resizeContainers(cpuLimit, currentPodCpuQuota, *podResources.CpuQuota); errUpdate != nil {
+				return
+			}
+		}
+	}
+
+	// Step 8: start containers in podContainerChanges.ContainersToStart.
 	for _, idx := range podContainerChanges.ContainersToStart {
 		start("container", containerStartSpec(&pod.Spec.Containers[idx]))
 	}
 
 	return
 }
 
+func (m *kubeGenericRuntimeManager) updatePodContainerResources(pod *v1.Pod, podStatus *kubecontainer.PodStatus, resourceName string, containersToUpdate []containerToUpdateInfo) error {
+	type updatedResourcesInfo struct {
+		cStatus *kubecontainer.Status
+		rLimits v1.ResourceList
+		rAllocs v1.ResourceList
+	}
+	updatedResourcesMap := make(map[kubecontainer.ContainerID]updatedResourcesInfo)
+
+	for _, cInfo := range containersToUpdate {
+		container := cInfo.apiContainer
+		containerStatus := cInfo.apiContainerStatus
+		specLimits := container.Resources.Limits
+		statusAllocs := containerStatus.ResourcesAllocated
+		defer func() {
+			container.Resources.Limits = specLimits
+			containerStatus.ResourcesAllocated = statusAllocs
+		}()
+		statusLimits := containerStatus.Resources.Limits
+		statusRequests := containerStatus.Resources.Requests
+		// Runtime container status resources, if set, takes precedence.
+		if cInfo.kubeContainerStatus.Resources.Limits != nil {
+			statusLimits = cInfo.kubeContainerStatus.Resources.Limits
+		}
+		if cInfo.kubeContainerStatus.Resources.Requests != nil {
+			statusRequests = cInfo.kubeContainerStatus.Resources.Requests
+		}
+		switch resourceName {
+		case cpuRequest:
+			container.Resources.Limits = v1.ResourceList{
+				v1.ResourceCPU:    *statusLimits.Cpu(),
+				v1.ResourceMemory: *statusLimits.Memory(),
+			}
+			containerStatus.ResourcesAllocated = v1.ResourceList{
+				v1.ResourceCPU:    *statusAllocs.Cpu(),
+				v1.ResourceMemory: *statusRequests.Memory(),
+			}
+		case cpuLimit:
+			container.Resources.Limits = v1.ResourceList{
+				v1.ResourceCPU:    *specLimits.Cpu(),
+				v1.ResourceMemory: *statusLimits.Memory(),
+			}
+			containerStatus.ResourcesAllocated = v1.ResourceList{
+				v1.ResourceCPU:    *statusRequests.Cpu(),
+				v1.ResourceMemory: *statusRequests.Memory(),
+			}
+		case memLimit:
+			container.Resources.Limits = v1.ResourceList{
+				v1.ResourceCPU:    *statusLimits.Cpu(),
+				v1.ResourceMemory: *specLimits.Memory(),
+			}
+			containerStatus.ResourcesAllocated = v1.ResourceList{
+				v1.ResourceCPU:    *statusRequests.Cpu(),
+				v1.ResourceMemory: *statusAllocs.Memory(),
+			}
+		}
+		if err := m.updateContainerResources(pod, container, cInfo.kubeContainerStatus.ID); err != nil {
+			klog.ErrorS(err, "updateContainerResources failed", "container", container.Name, "cID", cInfo.kubeContainerStatus.ID, "pod", format.Pod(pod))
+			return err
+		}
+		cInfo.apiContainerStatus.Resources.Limits = container.Resources.Limits
```

Comment:
This seems to be something we should do when generating api pod status based on the actual status, instead of updating it here.

Reply:
taken a TODO for this.